### PR TITLE
MKLDNN layout: Support for batch norm operator

### PR DESCRIPTION
**Pull request by mozga-intel**
Waiting for supports of the mkldnn’s layout. 

Please have a look at this batch norm operator supported by the MKLDNN’s layout and assess if we can keep this design of the code. 

This code uses the implementation of layout which is available in the last pull-request. Therefore some of the function in this code are not available in this pull-request, but are available here. 

This version of code can be merged into the main branch when the pull-request with layout is accepted. 
The concept of splits of the long code was proposed by @luotao1 

Pull-request is related to #11040

### CODE REVIEW

**Comment by mozga-intel** [comment]
@luotao1 The code is prepared to the code-review process.

File: paddle/fluid/operators/batch_norm_mkldnn_op.cc [hunk 1]
```diff
@@ -238,88 +228,132 @@ class BatchNormMKLDNNGradOpKernel : public paddle::framework::OpKernel<T> {
...
-                                   cast_const_to_void(diff_y->data<T>())};
+        batch_norm_bwd_desc, mkldnn_engine, *batch_norm_fwd_pd};
+
+    // reorder user_diff_dst if it's not in preferred format
+    auto diff_dst_memory = user_diff_dst_memory;
+    primitive reorder_diff_dst;
+    bool is_diff_dst_reordered = false;
+    if (diff_dst_pd != user_diff_dst_memory.get_primitive_desc()) {
+      diff_dst_memory = memory(diff_dst_pd);
+      reorder_diff_dst = reorder(user_diff_dst_memory, diff_dst_memory);
```
Comment (tensor-tang) [inline paddle/fluid/operators/batch_norm_mkldnn_op.cc]: Inplace reorder？

Comment (mozga-intel) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc]: @tensor-tang Yes.

Comment (tensor-tang) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc]: But does mkldnn reorder support inplace?

Comment (tensor-tang) [reply paddle/fluid/operators/batch_norm_mkldnn_op.cc]: Actually my point is does `mkldnn::reorder` support inplace, not `batchnorm`.
AFAIK, `mkldnn::reorder` does not support inplace yet.

While, you are not using inplace reorder here. 
 ```
auto diff_dst_memory = user_diff_dst_memory;
...
diff_dst_memory = memory(diff_dst_pd);
```

So that's fine.
